### PR TITLE
[cpp-package] Add monitor, optimizers, metrics and bug fixes

### PR DESCRIPTION
**Pull request by lx75249**
- Force MSVC to use UTF-8 (Recent doc updates contain UTF-8 characters, MSVC regards source files as system code page by default)
- cpp-package
  - supports MSVC and python 3
  - Add monitor interface
  - Add new optimizers (RMSProp, AdaGrad, AdaDelta, Adam)
  - Add new metrics (MAE, MSE, RMSE, PSNR)
  - Fix initializer bugs
  - Fix symbol loading bug

### CODE REVIEW

**Comment by lx75249** [comment]
@piiswrong This pr is ready for review. And any idea why cuda compilation failed in windows?

File: cpp-package/include/mxnet-cpp/monitor.hpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+      std::ostringstream out;
+      out << ndarray;
+      str = out.str();
+    }
+
+    LG << "Batch: " << std::get<0>(stat) << ' ' << std::get<1>(stat) << ' ' << str;
+  }
+}
+
+void Monitor::executor_callback(const char *name, NDArrayHandle handle,
```
Comment (sifmelcara) [inline cpp-package/include/mxnet-cpp/monitor.hpp]: miss an inline keyword.

File: cpp-package/include/mxnet-cpp/ndarray.h [hunk 2]
```diff
@@ -405,6 +407,8 @@ class NDArray {
  private:
   std::shared_ptr<NDBlob> blob_ptr_;
 };
+
+std::ostream& operator<<(std::ostream& out, const NDArray &ndarray);
```
Comment (sifmelcara) [inline cpp-package/include/mxnet-cpp/ndarray.h]: miss an inline keyword.

File: cpp-package/include/mxnet-cpp/optimizer.hpp [hunk 3]
```diff
@@ -12,14 +12,37 @@
...
+  return mxnet::cpp::Operator("sqrt")
+    .SetInput("data", data)
+    .Invoke()[0];
+}
+
+}  // namespace
+
 namespace mxnet {
 namespace cpp {
+Optimizer::Optimizer(unsigned begin_num_update)
```
Comment (sifmelcara) [inline cpp-package/include/mxnet-cpp/optimizer.hpp]: miss an inline keyword.

File: cpp-package/include/mxnet-cpp/optimizer.hpp [hunk 4]
```diff
@@ -132,6 +173,216 @@ inline void SGDOptimizer::CreateState_(int index, NDArray weight) {
...
+}
+
+inline void AdamOptimizer::CreateState_(int index, NDArray weight) {
+  mean_[index] = new NDArray(weight.GetShape(), weight.GetContext());
+  *mean_[index] = 0;
+  var_[index] = new NDArray(weight.GetShape(), weight.GetContext());
+  *var_[index] = 0;
+}
+
+AdaGradOptimizer::AdaGradOptimizer(unsigned begin_num_update)
```
Comment (sifmelcara) [inline cpp-package/include/mxnet-cpp/optimizer.hpp]: miss an inline keyword.

**Comment by Hebali** [comment]
Has progress on this stalled?

**Comment by lx75249** [comment]
ping @piiswrong

**Comment by piiswrong** [comment]
@lx75249 Please take a look at recent change to metrics.py and make sure this is compatible

**Comment by lx75249** [comment]
Existing apis in cpp version are compatible. As for new functions like `update_dict`, I will add them along with the module.

**Pull request merged**